### PR TITLE
util: extend row format with checksum

### PR DESCRIPTION
**Pull request by zyguan**
### What problem does this PR solve?

Issue Number: ref #42747

Problem Summary: see #42747 

### What is changed and how it works?

Add checksum-related utilities in tidb.
- add `Encoder.WithChecksum` to enable/disable checksum.
- add `Encoder.Checksum()` to calculate the checksum of input datums.
- add `Encoder.EncodeWithExtraChecksum()` to append extra checksum to the row value. (the extra checksum is calculated by `Encoder.Checksum()`)
- let row codec export `GetChecksum/GetExtraChecksum` to access the checksum info encoded in the row value.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: util/rowcodec/encoder.go [hunk 1]
```diff
@@ -34,29 +34,69 @@ type Encoder struct {
...
+	}
+	err := encoder.encodeDatums(sc, colIDs, values)
+	if err != nil {
+		return nil, err
+	}
+	return encoder.row.toBytes(buf[:0]), nil
+}
+
+// Checksum caclulates the checksum of datumns.
+func (encoder *Encoder) Checksum(sc *stmtctx.StatementContext, colIDs []int64, values []types.Datum) (uint32, error) {
```
[the remark this answers is not in the document]
Comment (zyguan) [reply util/rowcodec/encoder.go:65]: No, it used to calculate the extra checksum (the second checksum), and ticdc may also use it to calculate checksum of datums.

File: util/rowcodec/encoder.go [hunk 2]
```diff
@@ -34,29 +34,69 @@ type Encoder struct {
...
 	encoder.reset()
+	err := encoder.encodeDatums(sc, colIDs, values)
+	if err != nil {
+		return nil, err
+	}
+	return encoder.row.toBytes(buf[:0]), nil
+}
+
+// EncodeWithExtraChecksum likes Encode but also appends an extra checksum if checksum is enabled.
+func (encoder *Encoder) EncodeWithExtraChecksum(sc *stmtctx.StatementContext, colIDs []int64, values []types.Datum, checksum uint32, buf []byte) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (zyguan) [reply util/rowcodec/encoder.go:52]: Here we mean to append the extra checksum. The first checksum is automatically calculated as `crc32(encoder.row.data)` when `encoder.WithChecksum` is true.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Review by 3AceShowHand — APPROVED** [review]

**Comment by 3AceShowHand** [comment]
LGTM

**Pull request merged**